apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -295,590 +412,954 @@ poll(Node, ReturnAddr, Shard, Iterator, Opts = #{timeout := Timeout}) ->
+        subs = Subs
+    } = BeamMakerNode,
+    Flags0 =
+        case IsLagging of
+            true -> ?DISPATCH_FLAG_LAGGING;
+            false -> 0
         end,
-    %% Execute callbacks:
-    OnMatch(MatchReqs),
-    %% Split matched requests by destination node:
-    ReqsByNode = maps:groups_from_list(
-        fun(#poll_req{node = Node}) -> Node end,
-        fun(#poll_req{return_addr = RAddr, it = It}) ->
-            {RAddr, It}
+    Destinations =
+        maps:fold(
+            fun(SubId, #beam_builder_sub{s = SubS0, mask = MaskEncoder, n_msgs = NMsgs}, Acc) ->
+                #sub_state{client = Client, it = It0} = SubS0,
+                ?tp(beamformer_fulfilled, #{sub_id => SubId}),
+                %% Update the iterator, start key and message
+                %% matcher for the subscription:
+                {ok, It} = update_iterator(CBM, DBShard, It0, NextKey),
+                %% Update the subscription state:
+                SubS = SubS0#sub_state{it = It, start_key = NextKey},
+                ets:insert(SubTab, SubS),
+                %% Update sequence number and check if the
+                %% subscription should remain active:
+                {Active, SeqNo} = keep_and_seqno(SubTab, SubS0, NMsgs),
+                %% Update the worker's queue:
+                case Active of
+                    true ->
+                        %% Subscription is keeping up with the
+                        %% acks. Update the worker's active
+                        %% queue:
+                        Flags = Flags0,
+                        QueueUpdate(
+                            SubS0,
+                            SubS
+                        );
+                    false ->
+                        %% This subscription is not keeping up
+                        %% with the acks. Freeze the
+                        %% subscription by dropping it from
+                        %% the worker's active queue:
+                        Flags = Flags0 bor ?DISPATCH_FLAG_STUCK,
+                        disown_stuck_subscription(DBShard, SubTab, SubS),
+                        QueueDrop(SubS0)
+                end,
+                DispatchMask = emqx_ds_dispatch_mask:enc_finalize(BatchSize + 1, MaskEncoder),
+                [
+                    ?DESTINATION(Client, SubId, SeqNo, DispatchMask, Flags, It)
+                    | Acc
+                ]
+            end,
+            [],
+            Subs
+        ),
+    %% Send the beam to the destination node:
+    send_out(DBShard, Node, lists:reverse(PackRev), Destinations).
+
+send_out_final_term_to_node(DBShard, SubTab, Term, Node, Reqs) ->
+    Mask = emqx_ds_dispatch_mask:encode([true]),
+    Destinations = lists:map(
+        fun(
+            #sub_state{
+                client = Client, req_id = SubId, it = It, flowcontrol = FC
+            }
+        ) ->
+            {_MaxUnacked, ARef} = FC,
+            SeqNo = atomics:add_get(ARef, ?fc_idx_seqno, 1),
```

Comment:
Q: could this also race with `atomics:add_get(ARef, ?fc_idx_seqno, Nmsgs)` from `keep_and_seqno`?

Reply:
This function is called from the same process that calls beams_conclude -> keep_and_seqno.